### PR TITLE
Implement Blobstore Api

### PR DESCRIPTION
**Pull request by nabil-dbz**
This pull request implements the Blobstore Api:
- When the user opens the modal to update the profile picture, the UploadUrlServlet gets called to create an upload url
- When the user chooses the picture and confirms, the UploadFileServlet gets fetched to generate a blobKey
- The UploadFileServlet redirects to the UserServlet by putting the blobKey as a query string
- The UserServlet forwards the blobKey to the profile script which fetchs the ServeFileServlet to get the image and store it in the database

### CODE REVIEW

File: planet/src/main/java/com/google/planet/servlets/ServeFileServlet.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+
+@WebServlet("/serve")
```
Comment (EkkiD) [inline planet/src/main/java/com/google/planet/servlets/ServeFileServlet.java]: this name is at the top level for all of planet. In that context, "serve" isn't very descriptive. Maybe 'serveblobstore' or 'getimage' or 'serveimage', or something like that

Comment (EkkiD) [reply planet/src/main/java/com/google/planet/servlets/ServeFileServlet.java]: or "serve-file" to match your "upload-file" path below

File: planet/src/main/java/com/google/planet/servlets/UploadUrlServlet.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+@WebServlet("blobstore-upload-url")
+public class UploadUrlServlet extends HttpServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String username = request.getParameter("username");
+        BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+        String uploadUrl = blobstoreService.createUploadUrl("/upload-file?username=" + username);
```
Comment (EkkiD) [inline planet/src/main/java/com/google/planet/servlets/UploadUrlServlet.java:37]: nit: add a comment that the passed parameter will be called by blobstore after the upload is complete.

File: planet/src/main/java/com/google/planet/servlets/UploadFileServlet.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+
+@WebServlet("/upload-file")
+public class UploadFileServlet extends HttpServlet {
```
Comment (EkkiD) [inline planet/src/main/java/com/google/planet/servlets/UploadFileServlet.java:32]: nit: add a comment explaining how this is called (if I understand correctly, it's called by Blobstore after the user completes the upload?)

File: planet/src/main/webapp/js/profile.js [hunk 4]
```diff
@@ -100,30 +101,35 @@ function switchSection(linkId, sectionId) {
...
-    
-        if (user['emailVerified']) {
+        // If the blobkey exists, get the image url and store in the database
+        if (blobKey !== 'null') {
+            let picture = await fetch('/serve?blob-key=' + blobKey);
+            database.ref('/users/' + userId).update({
+                profilePic: picture.url
+            })
+            .then(() => {
+                window.location.href = '/user/' + username;
```
Comment (EkkiD) [inline planet/src/main/webapp/js/profile.js]: Why do we need to redirect after this DB write?

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: planet/src/main/webapp/js/profile.js (hunk 4 above)
Comment (nabil-dbz) [reply planet/src/main/webapp/js/profile.js]: Actually, it's just to hide the blobkey from the user. I don't know if it's important but if we don't redirect the user can see the blobkey from the profile's URL. Do you think I should remove it?

File: planet/src/main/java/com/google/planet/servlets/UploadFileServlet.java (hunk 3 above)
Comment (nabil-dbz) [reply planet/src/main/java/com/google/planet/servlets/UploadFileServlet.java:32]: Done!

File: planet/src/main/java/com/google/planet/servlets/UploadUrlServlet.java (hunk 2 above)
Comment (nabil-dbz) [reply planet/src/main/java/com/google/planet/servlets/UploadUrlServlet.java:37]: Done!

File: planet/src/main/java/com/google/planet/servlets/ServeFileServlet.java (hunk 1 above)
Comment (nabil-dbz) [reply planet/src/main/java/com/google/planet/servlets/ServeFileServlet.java]: Done!

**Review by zhoua0805 — APPROVED** [review]
LGTM

File: planet/src/main/webapp/js/profile.js (hunk 4 above)
Comment (EkkiD) [reply planet/src/main/webapp/js/profile.js]: What does the blob key give the user access to? Does it let them just download their own profile pic? If so, it should be fine to leave it.

Another consideration would be whether or not this is noticeable for the user. Does this line cause the whole page to reload or does the browser just change the URL bar but not reload anything?

Comment (nabil-dbz) [reply planet/src/main/webapp/js/profile.js]: It actually reloads the page that's why I didn't like this solution much. As for the info in the blobkey, it's just the blobkey of the user's profile picture. I will remove it then and keep the URL with the blobkey!

**Review by EkkiD — APPROVED** [review]
LGTM

**Pull request merged**